### ISSUE TITLE
No need for Error message with negative dataLabels: padding

### ISSUE THREAD

**Issue by drmrbrewer** [highcharts/highcharts#10267]
I'd struggled for ages to fine-tune the padding in my dataLabels... I was finding that the padding was too big and non-uniform for my requirements (space is really really limited and even a bit of extra label padding is wasted space).  Even trying to fiddle with it via setting explicit `padding` in the html css wasn't helping (can't set a negative css padding to pull back what is already too big).

Then I discovered that you can actually set a *negative* value for the `dataLabels: padding` option, which results in a much better (for me) starting point for the data labels boxes... it's too tight to the label text in places but it now allows me to fine tune and add *positive* css padding to get to what I want.

This is demonstrated in the following jsfiddle: https://jsfiddle.net/vy3n6zoq/1/

OK, so this works, so what is the issue?  Why I am posting here?

Reason is that I've now noticed in the browser console that the following highcharts errors are being output:

    Error: <rect> attribute width: A negative value is not valid.

The chart still renders fine, so the Error can't be fatal, but why is it issued in the first place?  It all seems to work OK with a negative padding value, and it is useful to be *able to* set a negative padding (not least for my own use case), so why the Error?  

I guess my concern is that in future some change will be made in highcharts which means that this Error is fatal, meaning that I'm no longer able to fine-tune my data labels as required.

Is it possible to change the source code to make setting a negative padding "legitimate", without any Error being issued?

**Comment by sebastianbochan** [comment]
Hi @drmrbrewer,
Thank you for reporting. In general padding has to be positive (or zero). It correct with general padding properties (CSS).

In your example Im not sure what you would like to achieve, because from one hand you try to reduce padding (by negative) but from another you add paddings in formatter (inline styles). 

When I reset padding (by set 0), you have as much datalabels as you need (according to your description).

**Demo:**
- https://jsfiddle.net/BlackLabel/x9t7cL02/

**Comment by drmrbrewer** [comment]
Hi @sebastianbochan.

Setting a negative initial padding allows me to reduce the box size right down to a better "starting point" for adding my own custom padding via inline styles... and note that my inline style padding is different top/bottom/left/right.  If I start with with zero padding, then the label padding is *already too big* in some directions and I cannot get the effect I want (can't apply negative inline padding).

> When I reset padding (by set 0), you have as much datalabels as you need

This is not the case.  On Chrome, for example, here is a clip from your example (zero padding):

![image](https://user-images.githubusercontent.com/9009792/53955384-cd1b3b80-40d0-11e9-9973-c4074afc84c4.png)

And here is what I can achieve with negative starting padding and then building out (fine tuning) from there via inline padding style:

![image](https://user-images.githubusercontent.com/9009792/53955448-f1771800-40d0-11e9-8d71-b93c19c748e7.png)

Notice that it has allowed me to create a much "neater" box that fits tighter to the label text.  In your example, there is too much padding at the bottom, and I can't correct for that via inline style.

So, being able to set a negative padding value is *useful* and should be retained, but it shouldn't result in any error messages (because it clearly works OK).

Note that I appreciate that this is browser dependent but my use case is within a controlled and known environment (either phantomjs or a known version of Chrome) so I can be exacting with my adjustments!

**Comment by sebastianbochan** [comment]
Hi @drmrbrewer,
Im thinking about different solution which should be more flexible. Please turn off default background and use css / inline styles in your custom div to set the background.

**Demo:**
- https://jsfiddle.net/BlackLabel/asgwn6k1/

Let me know if its better for you scenario.

**Comment by drmrbrewer** [comment]
I'm not sure how moving background inline helps?  I still cannot achieve what I want, which is to get a label box that fits more closely to the text.  

This is a clip from your latest example:

![image](https://user-images.githubusercontent.com/9009792/53957327-fab6b380-40d5-11e9-8009-3f7822497635.png)

The labels are clearly bigger than what I want (see my earlier clip).  It only seems possible to achieve that by setting a negative padding in highcharts dataLabels, and *also* applying an inline style on top of that.

My use case is for producing a chart in a very small area, where space is really at a premium, and having label boxes that are even a *little* bit too big is a problem.

In summary... I'm happy with the current ability to set a negative padding... I can achieve what I want.  My only concern is the Error message that results.  If that is non-fatal, and if the ability to set a negative padding in highcharts will be supported in future, I'm happy... I'll just ignore the Errors.

**Comment by pawelfus** [comment]
Hi @drmrbrewer 

Let me explain @sebastianbochan's solution a bit more: it's not about moving styles to inline, but replacing SVG's background with HTML styles. Since you use HTML to define dataLabels and you want very specific design (for me negative padding is peculiar) then let's leverage that HTML to style your datalabels. Maybe this example will be a bit better: https://jsfiddle.net/BlackLabel/asgwn6k1/5/

Simply add class to the element you want to style:
```javascript
        formatter: function() {
          // fine-tune the padding here...
          return '<div align="center" class="custom-datalabel">' + this.y + '</div>';
        }
```

Then you are free to change anything in CSS:
```css
.custom-datalabel {
  background-color: rgba(252, 255, 197, 0.7);
  border-radius: 5px;
  border: 1px solid #aaa;
  padding-left: 0px;
  padding-right: 0px;
  padding-bottom: 0px;
  padding-top: 0px;
  line-height: 8px;
}
```

Note: To decrease "padding" area in CSS, I simply used `line-height`. 

Let us know what do you think about this solution. I'm not sure what to do about negative padding. For me, throwing errors is wrong, but adding counterintuitive **negative** padding - it's worse. Solution for this issue is to validate padding value (to be positive or, when negative, rounded to 0), which certainly is not what you want..

**Comment by drmrbrewer** [comment]
Maybe it's font dependent, but I find that using `line-height` I can squish the label box vertically but for me the residual left/right "padding" in my own chart is more than in your example jsfiddle... i.e. I cannot get it tight to the text left and right.  I will experiment with this approach, if setting a negative padding is not something that will be "legal".

**Comment by pawelfus** [comment]
@drmrbrewer - just to be on the same page, you can keep fill/color logic in the `formatter` and use CSS classname only for the styles that don't change. I prefer changing classname instead of inline styles. Probably simple `document.getElementsByClassName('custom-datalabel');` could help with changing other styles.

**Comment by drmrbrewer** [comment]
@pawelfus indeed I am changing class names to change styles in some respects.  But at present I'm also inlining some styles, partly because I'm using a post-render method (as taught [here](https://www.highcharts.com/forum/viewtopic.php?f=9&t=38588)) to re-colour labels to match an underlying line that has a `linearGradient` line colour.  So I'm not sure that using a class is appropriate here to set colour of the labels, because the associated line has a colour that changes for each point.

Anyhow, I've now adapted my code based on your helpful example, i.e. moving styling from SVG to HTML.  Actually, it seems to be a lot neater as a result :)  I hadn't really appreciated previously that by using html for labels, and yet still setting the label box stuff via native highcharts, that I'd effectively end up with labels that were part-SVG and part-HTML.  That's right isn't it?  I can see this when rendered on a touch screen device, because zooming the chart would separate the label during the zoom into the box and the text (not great UI).  Now the (pure html) label stays all in one piece during a zoom.  And yes, using `line-height` I can make the border box very tight to the text, allowing me to adjust outwards using padding.

So, all is good, even without resorting to use of negative padding 👍

**Comment by pawelfus** [comment]
Great, thanks for the details! And yes, you are correct - `useHTML` + `borderSomething` creates HTML text and SVG background box.
